cmd/bosun/expr/annotate.go

```diff
@@ -0,0 +1,254 @@
+			continue
+		}
+		if aEnd.After(*reqEnd) {
+			aDurationBeforeReqEnd := reqEnd.Sub(aStart)
+			series[time.Unix(int64(i), 0).UTC()] = aDurationBeforeReqEnd.Seconds()
+		}
+	}
+	if len(series) == 0 {
+		series[time.Unix(0, 0).UTC()] = math.NaN()
+	}
+	return &Results{
+		Results: []*Result{
+			{Value: series},
+		},
+	}, nil
+}
+
+func AnCounts(e *State, T miniprofiler.Timer, filter, startDuration, endDuration string) (r *Results, err error) {
+	reqStart, reqEnd, err := procDuration(e, startDuration, endDuration)
+	if err != nil {
+		return nil, err
+	}
+	filteredAnnotations, err := getAndFilterAnnotations(e, reqStart, reqEnd, filter)
+	if err != nil {
+		return nil, err
+	}
+	series := make(Series)
+	for i, a := range filteredAnnotations {
+		aStart := a.StartDate.Time
+		aEnd := a.EndDate.Time
+		inBounds := (aStart.After(*reqStart) || aStart == *reqStart) && (aEnd.Before(*reqEnd) || aEnd == *reqEnd)
+		entirelyOutOfBounds := aStart.Before(*reqStart) && aEnd.After(*reqEnd)
+		if inBounds || entirelyOutOfBounds {
+			// time has no meaning here, so we just make the key an index since we don't have an array type
+			series[time.Unix(int64(i), 0).UTC()] = 1
+			continue
+		}
+		aDuration := aEnd.Sub(aStart)
+		if aDuration == 0 {
+			// This would mean an out of bounds. Should never be here, but if we don't return an error in the case that we do end up here then we might panic on divide by zero later in the code
+			return nil, fmt.Errorf("unexpected annotation with 0 duration outside of request bounds (please file an issue)")
+		}
+		if aStart.Before(*reqStart) {
+			aDurationAfterReqStart := aEnd.Sub(*reqStart)
+			percentBeforeStart := float64(aDurationAfterReqStart) / float64(aDuration)
+			series[time.Unix(int64(i), 0).UTC()] = percentBeforeStart
+			continue
+		}
+		if aEnd.After(*reqEnd) {
+			aDurationBeforeReqEnd := reqEnd.Sub(aStart)
+			percentAfterEnd := float64(aDurationBeforeReqEnd) / float64(aDuration)
+			series[time.Unix(int64(i), 0).UTC()] = percentAfterEnd
+		}
+	}
+	if len(series) == 0 {
+		series[time.Unix(0, 0).UTC()] = math.NaN()
+	}
+	return &Results{
+		Results: []*Result{
+			{Value: series},
+		},
+	}, nil
+}
+
+// AnTable returns a table response (meant for Grafana) of matching annotations based on the requested fields
+func AnTable(e *State, T miniprofiler.Timer, filter, fieldsCSV, startDuration, endDuration string) (r *Results, err error) {
+	start, end, err := procDuration(e, startDuration, endDuration)
+	if err != nil {
+		return nil, err
+	}
+	columns := strings.Split(fieldsCSV, ",")
+	columnLen := len(columns)
+	if columnLen == 0 {
+		return nil, fmt.Errorf("must specify at least one column")
+	}
+	columnIndex := make(map[string]int, columnLen)
+	for i, v := range columns {
+		// switch is so we fail before fetching annotations
+		switch v {
+		case "start", "end", "owner", "user", "host", "category", "url", "message", "duration", "link":
```

Comment:
kinda disliking this style. Could this be more data-driven? Thinking something like define a `map[string]func(annotation)interface{}` and all of the column checks and data fetching use that. Could kill the giant switch, and some of the duplication here. Maybe?